### PR TITLE
Add autologging safety utils to tensorflow

### PR DESCRIPTION
**Pull request by harupy**
## What changes are proposed in this pull request?

Add autologging safety utils to tensorflow.

## How is this patch tested?

Existing unit tests.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tensorflow.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
 from mlflow.utils.file_utils import _copy_file_or_tree
```
Comment (harupy) [inline mlflow/tensorflow.py:37]: In tensorflow autologging, `_manage_active_run` and `_AUTOLOG_RUN_ID` manage an active run across the patched functions.

https://github.com/mlflow/mlflow/blob/d1abdc2ee492958c7376db43c0be4780d403c380/mlflow/tensorflow.py#L854-L900

Comment (harupy) [reply mlflow/tensorflow.py:36]: @dbczumar Is it possible to patch multple functions with `safe_patch` and share the same active run across them?

Comment (harupy) [reply mlflow/tensorflow.py:36]: ↑ Just never mind

Comment (harupy) [reply mlflow/tensorflow.py:36]: The current tensorfolw autologging works like this (maybe I'm wrong):

```python
# no active run exists

classifier.train(input_fn=lambda: input_fn(train, train_y, training=True), steps=training_steps)
# -> creates a new run, store the run_id in `_AUTOLOG_RUN_ID`, and terminates it

classifier.export_saved_model(directory, receiver_fn)
# -> Restart the run using `_AUTOLOG_RUN_ID` and save the trained model in it
```

File: mlflow/tensorflow.py [hunk 2]
```diff
@@ -989,110 +985,118 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
...
-                    kwargs["callbacks"], log_dir = _setup_callbacks(
-                        kwargs["callbacks"], log_models, metrics_logger
-                    )
-                else:
-                    kwargs["callbacks"], log_dir = _setup_callbacks([], log_models, metrics_logger)
+    def fit(original, self, *args, **kwargs):
+        active_run = mlflow.active_run()
+        if MLFLOW_AUTOLOGGING in active_run.data.tags:
+            global _AUTOLOG_RUN_ID
+            _AUTOLOG_RUN_ID = active_run.info.run_id
```
Comment (harupy) [inline mlflow/tensorflow.py:992]: If the current active run was generated by autologging, store the run_id in `global _AUTOLOG_RUN_ID`.

File: mlflow/tensorflow.py [hunk 3]
```diff
@@ -851,48 +858,37 @@ def autolog(every_n_iter=100, log_models=True):
...
-    def export_saved_model(self, *args, **kwargs):
+    def export_saved_model(original, self, *args, **kwargs):
         auto_end = False
         if not mlflow.active_run():
             global _AUTOLOG_RUN_ID
             if _AUTOLOG_RUN_ID:
                 try_mlflow_log(mlflow.start_run, _AUTOLOG_RUN_ID)
             else:
                 try_mlflow_log(mlflow.start_run)
+                try_mlflow_log(mlflow.set_tag, MLFLOW_AUTOLOGGING, FLAVOR_NAME)
```
Comment (harupy) [inline mlflow/tensorflow.py:889]: Manually set an `MLFLOW_AUTOLOGGING` tag.

File: mlflow/tensorflow.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/tensorflow.py:992]: What happens if a user creates their own run and performs autologging within it? Looks like we won't capture model artifacts in that case. I think this should just be `_AUTOLOG_RUN_ID = mlflow.active_run().info.run_id`

File: mlflow/tensorflow.py [hunk 4]
```diff
@@ -989,110 +985,118 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
...
-            _flush_queue()
-            _log_artifacts_with_warning(
-                local_dir=log_dir.location, artifact_path="tensorboard_logs"
-            )
-            if log_dir.is_temp:
-                shutil.rmtree(log_dir.location)
+        _flush_queue()
+        _log_artifacts_with_warning(local_dir=log_dir.location, artifact_path="tensorboard_logs")
+        if log_dir.is_temp:
+            shutil.rmtree(log_dir.location)
```
Comment (dbczumar) [inline mlflow/tensorflow.py:1026]: can we leverage `mlflow.utils.autologging_utils.PatchFunction.on_exception` here to ensure that the logdir is removed if an unexpected exception occurs? see https://github.com/dbczumar/mlflow/blob/da54e042ef52c4a5c4314bf320bcb036ebdf2b1e/mlflow/tensorflow.py#L992-L1045 for an example.

File: mlflow/tensorflow.py [hunk 5]
```diff
@@ -989,110 +985,118 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
...
+                kwargs["callbacks"], log_dir = _setup_callbacks(
+                    kwargs["callbacks"], log_models, metrics_logger
+                )
+            else:
+                kwargs["callbacks"], log_dir = _setup_callbacks([], log_models, metrics_logger)
+            result = original(self, *args, **kwargs)
+
+        _flush_queue()
+        _log_artifacts_with_warning(local_dir=log_dir.location, artifact_path="tensorboard_logs")
+        if log_dir.is_temp:
```
Comment (dbczumar) [inline mlflow/tensorflow.py:1064]: Same as https://github.com/mlflow/mlflow/pull/3861/files#r545584715 - can we leverage `mlflow.utils.autologging_utils.PatchFunction.on_exception`?

File: mlflow/tensorflow.py (hunk 2 above)
Comment (harupy) [reply mlflow/tensorflow.py:992]: If a user runs:

```
with mlflow.start_run():
    model.train(...)
    model.export_saved_model()
```

`export_saved_model` works like:

```python
    def export_saved_model(original, self, *args, **kwargs):
        auto_end = False
        if not mlflow.active_run():  # not mlflow.active_run() returns False
            global _AUTOLOG_RUN_ID
            if _AUTOLOG_RUN_ID:
                try_mlflow_log(mlflow.start_run, _AUTOLOG_RUN_ID)
            else:
                try_mlflow_log(mlflow.start_run)
                try_mlflow_log(mlflow.set_tag, MLFLOW_AUTOLOGGING, FLAVOR_NAME)
                auto_end = True

        serialized = original(self, *args, **kwargs)
        try_mlflow_log(...)  # logs the model in the active run

        if (
            mlflow.active_run() is not None and
            mlflow.active_run().info.run_id == _AUTOLOG_RUN_ID  # this return False
        ) or auto_end:
            try_mlflow_log(mlflow.end_run)
        return serialized
```

File: mlflow/tensorflow.py (hunk 5 above)
Comment (harupy) [reply mlflow/tensorflow.py]: Done!

File: mlflow/tensorflow.py (hunk 4 above)
Comment (harupy) [reply mlflow/tensorflow.py]: done!

File: mlflow/tensorflow.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/tensorflow.py]: Ah, got it. Totally missed the `if not mlflow.active_run()`. All good :)! Thanks for clarifying!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Pull request merged**